drivers/adc/mcp356x/mcp356x.c

```diff
@@ -0,0 +1,542 @@
+  uint32_t sample_index = 0;
+
+  for (uint8_t i = 0; i < MAX_CHANNELS; i++) {
+    uint8_t channel_id = i;
+    if (((sequence->channels >> channel_id) & 0b1) == 0) {
+      continue;
+    }
+
+    struct channel_map_entry cfg = data->channel_map[channel_id];
+    uint8_t config0 = (data->config0 & 0b01111111) | (1 << 7);
+    if (config0 != data->config0) {
+      (void)mcp_write_reg_8(config, MCP_Reg_CONFIG0, config0);
+      data->config0 = config0;
+    }
+
+    uint8_t config2 = (data->config2 & 0b11000111) | (cfg.gain_bits << 3);
+    if (config2 != data->config2) {
+      (void)mcp_write_reg_8(config, MCP_Reg_CONFIG2, config2);
+      data->config2 = config2;
+    }
+
+    uint8_t mux_reg = data->channel_map[i].mux_reg;
+    if (mux_reg != data->mux) {
+      (void)mcp_write_reg_8(config, MCP_Reg_MUX, mux_reg);
+      data->mux = mux_reg;
+
+      // Wait for result to come in after ADC reset because of MUX change
+      // right one
+      k_usleep(10);
+    }
+
+    uint32_t val = 12345;
+    res = mcp_read_reg_24(config, MCP_Reg_ADCDATA, &val);
+    if (res != 0) {
+      return res;
+    }
+    sample_array[sample_index] = val; // TODO FIX THIS
+    sample_index++;
+  }
+
+  return 0;
+}
+
+int mcp356x_channel_setup(const struct device *dev,
+                          const struct adc_channel_cfg *channel_cfg) {
+  if (!device_is_ready(dev)) {
+    LOG_ERR("Base mcp356x device is not setup. Can not setup channel");
+    return -ENODEV;
+  }
+  struct mcp356x_data *data = dev->data;
+
+  if (channel_cfg->channel_id > 8) {
+    LOG_ERR(
+        "register/channel %d of %s invalid. Any mcp356x device supports max 4 "
+        "differential channels or 8 single ended channels. You probably "
+        "didn't mean to have this many channels",
+        channel_cfg->channel_id, dev->name);
+    return -ENODEV;
+  }
+
+  uint8_t mux_vin_p = channel_cfg->input_positive;
+  if (mux_vin_p > 15) {
+    LOG_ERR("zephyr,positive invalid for %s channel %d.", dev->name,
+            channel_cfg->channel_id);
+  }
+  uint8_t mux_vin_m = channel_cfg->input_negative;
+  if (mux_vin_m > 15) {
+    LOG_ERR("zephyr,negative invalid for %s channel %d.", dev->name,
+            channel_cfg->channel_id);
+  }
+
+  uint8_t mux_reg = (mux_vin_p << 4) | mux_vin_m;
+  bool differential = channel_cfg->differential;
+  uint8_t gain_bits = 0;
+  switch (channel_cfg->gain) {
+  case ADC_GAIN_1_3:
+    gain_bits = 0b000;
+    break;
+  case ADC_GAIN_1:
+    gain_bits = 0b001;
+    break;
+  case ADC_GAIN_2:
+    gain_bits = 0b010;
+    break;
+  case ADC_GAIN_4:
+    gain_bits = 0b011;
+    break;
+  case ADC_GAIN_8:
+    gain_bits = 0b100;
+    break;
+  case ADC_GAIN_16:
+    gain_bits = 0b101;
+    break;
+  case ADC_GAIN_32:
+    gain_bits = 0b110;
+    break;
+  case ADC_GAIN_64:
+    gain_bits = 0b111;
+    break;
+  default:
+    LOG_ERR("unsupported channel gain '%d'", channel_cfg->gain);
+    return -ENOTSUP;
+  }
```

Comment:
Formatting. Also why not make an enum for the gain bits and assign to the variable

Reply:
The ADC_GAIN enum is built in to zephyr unlike the other enums so we don't have control over their underlying value. So, for this one it must be a switch.